### ISSUE TITLE
Help with error: "Trying to load a registered polymorphic type with an unregistered polymorphic cast."

### ISSUE THREAD

**Issue by GrahamZiervogel** [USCiLab/cereal#678]
Hello,

I haven't worked in C++ for a while and I am struggling to get polymorphic serialisation working for a small project. I am using Catch as a testing framework and have written some test cases for the project. One of these tests illustrates the problem I am having with cereal - namely that I cannot successfully use the load_and_construct method to create an object that has been serialized to .xml.

I am hoping that I have overlooked something very simple. I saw a number of github issues that were closely related to my problem, but none of the suggested fixes seem to work for me.

I will include my source code here. If anyone is able to take a look at it and provide feedback it would be greatly appreciated. I can also respond and provide additional information or context where required.

I believe the most relevant lines of code will be lines 33-70 of HeavyPosition.h. There you can also see a link to the most closely related issue on github I could find as well as my commented out attempt to replicate the solution from that issue.

Thanks, 

Graham

*Note: I did not include a bloated header I use called "HashDefinesForCpp.h". This is only used for a call to "to_string()" on line 21 of Position.cpp and so if one wants to compile the project this can be removed.

[cereal_polymorphic_problem.zip](https://github.com/USCiLab/cereal/files/5919712/cereal_polymorphic_problem.zip)
![tests](https://user-images.githubusercontent.com/26275872/106772404-e23fd280-6637-11eb-99c0-60706d173092.png)

**Comment by GrahamZiervogel** [comment]
For completeness, I attach the Catch files I am using as well.
[catch.zip](https://github.com/USCiLab/cereal/files/5919756/catch.zip)

**Comment by Bawz** [comment]
Hey, I took a quick look at your serialization method `inline void HeavyPosition::serialize(...)`. Maybe add method there to serialize base class recursively by adding line `ar( cereal::base_class<Position>(this)`. Polymorphic docs: https://uscilab.github.io/cereal/polymorphism.html

**Comment by GrahamZiervogel** [comment]
Thanks for the feedback.

Yes I did read the docs and went with the header implementation approached. I tried using cereal::base_class in both my serialize and my load_and_construct method, but then got a different error that I will try to reproduce now (I think it was an ambiguous call to overloaded constructor).

In general, "save" or the serialize method works well. But I run into problems with the "load" part, even if I try to specify default constructors.

**Comment by GrahamZiervogel** [comment]
Modifying the code to make use of cereal::base_class in both serialize and load_and_construct methods:

![1](https://user-images.githubusercontent.com/26275872/107018967-04049b00-6799-11eb-92f5-03d3f87f127b.PNG)

Leads to the following error at build time:

![0](https://user-images.githubusercontent.com/26275872/107018974-06ff8b80-6799-11eb-9f60-f032cca1ccc9.PNG)

**Comment by GrahamZiervogel** [comment]
Thanks very much for taking the time to help.

I will attempt to replicate your solution soon.

The code is a small extract from a larger project I was working on where I was using load_and_construct to allow me to make shared pointers to the objects and not have default constructors. 

I should hopefully be able to work around this and implement your suggestions.

**Comment by GrahamZiervogel** [comment]
Unfortunately this does not resolve the problem. I probably should have been more clear that the test case that is failing is the saving and loading of a *HeavyFund* object (which has a vector of HeavyPosition objects as a member):

![0](https://user-images.githubusercontent.com/26275872/107043166-86518700-67ba-11eb-9726-9680f4cc75d5.PNG)

The main file you include actually tests something that was passing in one of my original test cases - namely saving and loading of HeavyPositions.

The problem comes in with the HeavyPositions that are in a HeavyFund object.

A Fund has a vector of Positions.
A HeavyFund is a Fund.
A HeavyPosition is a Position.
A HeavyFund has a vector of HeavyPositions (which are Positions and therefore satisfy the requirement of the parent class).

I can successfully save and load Fund, Position, and HeavyPosition objects. But for some reason when trying to load HeavyFund objects I get the unregistered polymorphic cast error (which refers to HeavyPosition).

I have tried various combinations of cereal::base_class and CEREAL_REGISTER_POLYMORPHIC_RELATION with no success. 
(Note the above image was made using the source code from your zip file).

**Comment by GrahamZiervogel** [comment]
The error also seems to be self-referential (note that it says it cannot find path to base class HeavyPosition for type HeavyPosition, as if HeavyPosition is somehow being seen as its own base class).

**Comment by GrahamZiervogel** [comment]
OK this is quite bizarre...after noticing the above self-referential error message above I tried adding:

`CEREAL_REGISTER_POLYMORPHIC_RELATION(itshokholethi::HeavyPosition, itshokholethi::HeavyPosition)
`

thinking it could never work - but it did.

After adding this line all tests now pass. I am pretty sure this cannot be intended behaviour.

**Comment by Bawz** [comment]
Yeah I don't think `CEREAL_REGISTER_POLYMORPHIC_RELATION(itshokholethi::HeavyPosition, itshokholethi::HeavyPosition)` is intended behavior either.

**Comment by Bawz** [comment]
Are this lines present? 
Fund.h `CEREAL_REGISTER_TYPE(itshokholethi::Fund)`
HeavyFund.h `CEREAL_REGISTER_TYPE(itshokholethi::Fund)` and `CEREAL_REGISTER_POLYMORPHIC_RELATION(itshokholethi::Fund, itshokholethi::HeavyFund)`

**Comment by GrahamZiervogel** [comment]
I did not have `CEREAL_REGISTER_TYPE(itshokholethi::Fund)` in Fund.h but I added the other 2 this morning.

I will try register Fund now and see if that removes the need for the weird self-referential relation above.

**Comment by GrahamZiervogel** [comment]
No the above did not work. I also added `CEREAL_REGISTER_TYPE(itshokholethi::Position)` in Position.h as well for good measure.

The only working solution I have found is to include this weird `CEREAL_REGISTER_POLYMORPHIC_RELATION(itshokholethi::HeavyPosition, itshokholethi::HeavyPosition)`

Every other attempted solution gives error: "Trying to load a registered polymorphic type with an unregistered polymorphic cast. Could not find a path to base class (class itshokholethi::HeavyPosition) for type: class itshokholethi::HeavvPosition

**Comment by Bawz** [comment]
Okay because your Base classes are not pure virtual (abstract). I would now try to add `cereal::base_class<Base>(this)` too `archive(cereal::base_class<Fund>(this), ...)` for example like in documentation: https://uscilab.github.io/cereal/polymorphism.html#ensuring-a-path-exists-from-derived-to-base-type

This is really tricky.

**Comment by Bawz** [comment]
If all not working, try to reproduce on a very small project.

**Comment by GrahamZiervogel** [comment]
Thanks Bawz

I got a working solution for my small project here (albeit using the weird self-referential relation). I don't really understand what is causing cereal to see HeavyPosition as a derived class of itself (which is of course impossible), but registering this relation does seem to work. I saved and loaded a bunch of HeavyFunds with HeavyPositions and didn't have any problems. 

That being said, it would be nice to try to remove  `CEREAL_REGISTER_POLYMORPHIC_RELATION(itshokholethi::HeavyPosition, itshokholethi::HeavyPosition)` if I could.

I might try write an entirely new project with the same classes using the `cereal::base_class` calls in both HeavyFund and HeavyPosition at some point to see if this works. 

I did do some experimentation with this reading the documentation you highlight earlier this week and it didn't work then, but maybe I had some other problem. Worth a shot. I will report back here if I find anything (but it will only be some time next week).

**Comment by Bawz** [comment]
Yeah, give it a try! I will read it if you got further informations.
